## FIT 1043 Introduction to Data Science


###  Assignment 2 
### Akira Sato (33555052)

# Introduction 
This report includes the codes for the development of machine learning model specifically in this case Support Vector Machine, that would predict the credit score of people by utilising other given information in a file called FIT1043-Credit-Scores-Dataset.csv

This report is divided into few sctions as follows:
1. Introduction
2. Supervised Learning
3. Classification
4. Kaggle Submission (Prediction)
5. Conclusion



In [34]:
#importing necessary libraries
import numpy as np
import matplotlib.pyplot as pl
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,  f1_score, confusion_matrix, cohen_kappa_score 
from sklearn.preprocessing import MinMaxScaler
from sklearn import svm


#Reading file
c_score = pd.read_csv("FIT1043-Credit-Scores-Dataset.csv") 
c_score.dtypes

ID                            int64
Month                         int64
Age                           int64
Occupation                    int64
Annual_Income               float64
Monthly_Inhand_Salary       float64
Num_Bank_Accounts             int64
Num_Credit_Card               int64
Interest_Rate                 int64
Num_of_Loan                   int64
Delay_from_due_date           int64
Num_of_Delayed_Payment        int64
Changed_Credit_Limit        float64
Num_Credit_Inquiries          int64
Credit_Mix                    int64
Outstanding_Debt            float64
Credit_Utilization_Ratio    float64
Credit_History_Age            int64
Payment_of_Min_Amount         int64
Total_EMI_per_month         float64
Amount_invested_monthly     float64
Payment_Behaviour             int64
Monthly_Balance             float64
Credit_Score                  int64
dtype: object

In [35]:
c_score.head()

ID  Month  Age  Occupation  Annual_Income  Monthly_Inhand_Salary  \
0  106981      8   41           2      14619.585            1005.298750   
1  108774      1   28          12      70883.440            5663.953333   
2  111896      3   29          12      14395.830            1027.652500   
3   32731      2   25           1      11189.065            1159.422083   
4  128760      7   37           3      78956.730            6523.727500   

   Num_Bank_Accounts  Num_Credit_Card  Interest_Rate  Num_of_Loan  ...  \
0                  7                7             19            1  ...   
1                  4                4             10            3  ...   
2                  8                8             28            7  ...   
3                  6                3             15            3  ...   
4                  7                3             14            2  ...   

   Credit_Mix  Outstanding_Debt  Credit_Utilization_Ratio  Credit_History_Age  \
0           2            125.33                 38.883189                 222   
1           2            604.77                 31.131854                 356   
2           1           2841.00                 37.587389                  27   
3           2            761.18                 33.980973                 126   
4           2            436.82                 27.684657                 265   

   Payment_of_Min_Amount  Total_EMI_per_month  Amount_invested_monthly  \
0                      1            11.620889                32.846250   
1                      0            97.133997                39.858686   
2                      1            74.795382                31.947738   
3                      0            18.439801                16.806258   
4                      1           128.558654                70.788144   

   Payment_Behaviour  Monthly_Balance  Credit_Score  
0                202       279.724565             2  
1                201       526.033197             2  
2                201       258.713002             1  
3                201       324.284100             2  
4                102       669.025667             2  

[5 rows x 24 columns]

In [36]:
c_score.describe()

ID        Month          Age   Occupation  Annual_Income  \
count    2100.000000  2100.000000  2100.000000  2100.000000    2100.000000   
mean    83333.452381     4.411429    33.197143     7.927619   51244.805155   
std     41013.880408     2.279068    10.754180     4.325104   38801.235929   
min     10033.000000     1.000000    14.000000     1.000000    7011.685000   
25%     48108.000000     2.000000    24.000000     4.000000   19696.500000   
50%     81490.000000     4.000000    33.000000     8.000000   37087.920000   
75%    118895.250000     6.000000    42.000000    12.000000   73327.380000   
max    155610.000000     8.000000    56.000000    15.000000  179987.280000   

       Monthly_Inhand_Salary  Num_Bank_Accounts  Num_Credit_Card  \
count            2100.000000        2100.000000      2100.000000   
mean             4255.554221           5.407143         5.576667   
std              3226.388893           2.553747         2.099327   
min               319.556250           0.000000         0.000000   
25%              1641.972917           4.000000         4.000000   
50%              3101.737500           6.000000         5.000000   
75%              6118.762500           7.000000         7.000000   
max             15101.940000          11.000000        11.000000   

       Interest_Rate  Num_of_Loan  ...   Credit_Mix  Outstanding_Debt  \
count    2100.000000   2100.00000  ...  2100.000000       2100.000000   
mean       14.532857      3.50619  ...     2.057619       1444.016295   
std         8.736433      2.47685  ...     0.731132       1186.545886   
min         1.000000      0.00000  ...     1.000000          0.230000   
25%         7.000000      2.00000  ...     2.000000        566.605000   
50%        13.000000      3.00000  ...     2.000000       1149.035000   
75%        20.000000      5.00000  ...     3.000000       2001.147500   
max        34.000000      9.00000  ...     3.000000       4997.100000   

       Credit_Utilization_Ratio  Credit_History_Age  Payment_of_Min_Amount  \
count               2100.000000         2100.000000            2100.000000   
mean                  32.379813          221.545714               0.522381   
std                    5.199365          100.110503               0.499618   
min                   21.359116            4.000000               0.000000   
25%                   28.059563          142.000000               0.000000   
50%                   32.402470          219.000000               1.000000   
75%                   36.641141          303.000000               1.000000   
max                   46.724651          403.000000               1.000000   

       Total_EMI_per_month  Amount_invested_monthly  Payment_Behaviour  \
count          2100.000000              2100.000000        2100.000000   
mean            110.411491                56.302384         153.554762   
std             138.536282                40.246618          49.702836   
min               0.000000                 0.000000         101.000000   
25%              27.488358                27.842413         102.000000   
50%              67.878526                46.803308         201.000000   
75%             146.882432                73.229307         201.000000   
max            1779.103254               297.064670         203.000000   

       Monthly_Balance  Credit_Score  
count      2100.000000   2100.000000  
mean        399.349997      1.865714  
std         203.955200      0.674920  
min           7.438257      1.000000  
25%         269.255129      1.000000  
50%         337.412112      2.000000  
75%         473.138717      2.000000  
max        1181.113695      3.000000  

[8 rows x 24 columns]

Description of the provided data:
 As shown above, there are 24 different types of numerical data are provided, with all being either integer or float data type. There are 2100 data sets provided in total. The first few columns are general personal information such as id and ae etc, yet financial information are stored column 4 onwards. There seems to be in need of normalisation as the range of values in each column varies a lot.

# 2.Supervised learning

Supervised machine learning is a type of machine learning which learns its algorithm by the provided labelled data to make predictions and classification with a new data set. Supervised learning can be used for a broad range of tasks, such as classification and regression. Things that can be done using this type of machine learning would be spam detection and image classification.

Labelled data contains the raw data and its corresponding output such as features enabling the the data to be more clearly and consistently classified. This allows the algorithm to deduce the pattern more easily when the data is fed.

The training dataset is used to train the machine learning model. It consists of a set of labelled examples, where each example contains input data and its corresponding output label. The model is trained on this data by adjusting its parameters until it can accurately predict the output for the input data in the training set.

Once the model has been trained, it is tested on a diffrent unused dataset called the test dataset. The test dataset is used to evaluate how well the model can generalize to new, unseen data. The test dataset also consists of labelled examples but has never been used during the training process.

In [37]:
#Splitting features and labels
X = c_score.iloc[:,5:23].values # Input Data: Features
# X = c_score.iloc[:,[5,6,7,9,10,11,12,13,14,15,16,17,18,19,20,21,22]].values
# X = c_score.drop(('Credit_Score'), axis=1).copy()
y = c_score.iloc[:,23].values # Labeled Data: Credit score
print(y)

[2 2 1 ... 2 2 2]


In [38]:
#splitting the test and train data
X_train, X_test, y_train, y_test = train_test_split(
 X, y, test_size = 0.25, random_state = 0
)

# 3.Classification


Binary classification is a type of classification which devides the data into two classes. It can be used for things like differention of spam and non spam emails

On the other hand, multi-class classification separates data into three or more classes. For example, a multi-class classification task could involve classifying images of different types of plants based on its features.


In [39]:
# Data normalisation
sc = MinMaxScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

#The purpose of normalisation

The normalisation is required as the range of value in data varies significantly. for example, without normalisation, the data column that usually has a velue of around 10000 and the other data group which its value is a single digit. The former data has significantly more efffect on the output produced. Hence, it needs to be made into similar magnitude by normalisation.

#SVM Classification Model

 Support Vector Machine (SVM) is a type of supervised learning algorithm used for classification. This algorithm tries to find the hyperplane in the high-dimensional space that best separates the different classes. The hyperplane is chosen such that the margin (the space between the hyperplane and support vectors) between the two classes is maximized. This is different from linear regression where it is used to predict the output of continuous variable with input, this algorithm is used classify categorical data.
 
#What is Kernel?

 A kernel function in SVM is a mathematical function that allows SVM to find hyperplanes for complex calculation by providing shortcuts to execute simpler calculations. The kernel used in this model is polynomial. 

In [40]:
#SVM Model building
clf = svm.SVC(decision_function_shape='ovo',kernel = 'poly')
clf.fit(X_train, y_train)

SVC(decision_function_shape='ovo', kernel='poly')

In [41]:
clf

SVC(decision_function_shape='ovo', kernel='poly')

In [42]:
y_pred = clf.predict(X_test)

In [43]:
cm = confusion_matrix(y_test, y_pred)
cm

array([[ 83,  57,  14],
       [ 33, 216,  33],
       [  4,  30,  55]])

#Confusion matrix

A confusion matrix is a table that evaluates the performace of the machine learning model by comparing the predicted and true label of the datasets.

In 3x3 confusion matrix, the correct prediction would be the digonal data starting from the top left. Hence, the accuracy score for it would be (83+216+55)/(83+33+4+57+216+30+14+33+55) = 67.43

This tells that the built model has successfully predicted 67.43% of the datasets. which is considered to be okay.

In [44]:
#for double checking purpose
y_accuracy = accuracy_score(y_test, y_pred)
print('Accuracy (poly Kernel): ', "%.2f" % (y_accuracy*100))

Accuracy (poly Kernel):  67.43


What is QWK?

 QWK is a statistical measure of a performance of the model indicating how well is the agreement of the outputs, taking into account of possibility of agreement that occured by chance. The score is measured between 1,-1. And in general, the positive scores suggest good performance of the model.

In [45]:
qwk1 = cohen_kappa_score(y_test, y_pred)
print("Poly" + str(qwk1))

Poly0.4479495268138802


This score is considered to be moderate, which indicates that the model classifies the data better than solely being dependent on luck.

# 4. Kaggle Submission

In [46]:
#Reading competition data file
comp_c_score = pd.read_csv('FIT1043-Credit-Scores-Submission.csv')

In [47]:
comp_c_score.head()

ID  Month  Age  Occupation  Annual_Income  Monthly_Inhand_Salary  \
0  89760      7   17           1       60335.25              5120.9375   
1  44011      2   33           2       16706.97              1364.2475   
2  54153      4   37           3       32470.08              2733.8400   
3  25314      1   43           1       85324.44              7286.3700   
4  58305      4   29           4       93169.92              7794.1600   

   Num_Bank_Accounts  Num_Credit_Card  Interest_Rate  Num_of_Loan  ...  \
0                  9                9             15            2  ...   
1                  6                7             28            7  ...   
2                  6                6             19            2  ...   
3                  5                7             18            3  ...   
4                  6                7             18            2  ...   

   Num_Credit_Inquiries  Credit_Mix  Outstanding_Debt  \
0                     7           1           1814.22   
1                     9           1           3987.25   
2                     7           2           1456.87   
3                     3           2            713.09   
4                     1           2           1110.09   

   Credit_Utilization_Ratio  Credit_History_Age  Payment_of_Min_Amount  \
0                 37.592920                 206                      1   
1                 36.053673                 129                      1   
2                 23.567603                 206                      1   
3                 31.034484                 335                      0   
4                 26.212077                 322                      1   

   Total_EMI_per_month  Amount_invested_monthly  Payment_Behaviour  \
0            92.425471                91.839285                103   
1            70.367486                19.019972                201   
2            51.987693                63.583075                201   
3           197.305281               105.583784                202   
4           112.340298                63.048711                103   

   Monthly_Balance  
0       529.876377  
1       282.438875  
2       193.820905  
3       186.457750  
4       620.518394  

[5 rows x 23 columns]

In [48]:
X = comp_c_score.iloc[:,5:23].values
X = sc.fit_transform(X)
y_pred = clf.predict(X)

df = pd.DataFrame({'ID': comp_c_score.ID , 'Credit_Score': y_pred})

# Save to a CSV file
df.to_csv('99999999-AkiraSato-v1.csv', index=False)


# 5.Conclusion

 This assignment developed a SVM model using poly kernel which successfully passed a benchmark score given in the Kaggle by scoring 0.58839. The model takes in all the data from monthly_inhand_salary up to the last column. In other words, basic personal infrormation such as month and age etc were not considered to prouce a decent prediction of credit score.